[MRG] Pandas Interoperability section

fdroessler opened the pull request:
This PR aims to implement bits of #10453 as part of the python-sprints London doc sprint.

More details such as which bits will be worked on and by whom will follow as part of the sprint.

Commit:
initial commit

dirkesquire committed:
Added section on sparse conversion to pandas interoperability #10453

rth commented:
Do you plan to continue working on this @fdroessler ?

fdroessler commented:
@rth Yes, sorry it has been very busy. I will work on this over the next week, but if you want to contribute as well please let me know as there is a lot that can be done.

fdroessler committed:
Merge pull request #1 from dirkesquire/pandas_userguide

Added section on sparse conversion to pandas interoperability #10453

fdroessler committed:
Added basic content for different sections in Pandas Userguide.

fdroessler committed:
Merge branch 'pandas_userguide' of https://github.com/fdroessler/scikit-learn into pandas_userguide

fdroessler committed:
added some links and further examples.

fdroessler committed:
Merge branch 'master' into pandas_userguide

fdroessler commented:
@rth I have pushed an update-to-date version of this section. It would be good to know, if this in general fits what you (others) were thinking about for this section and get some initial feedback. There is of course plenty of things to add and fix but I wanted to get some initial feedback before I go off in the wrong direction.

rth commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,202 @@
+dia_matrix  Sparse matrix with diagonal storage
+dok_matrix  Dictionary of Keys based sparse matrix
+lil_matrix  Row-based linked list sparse matrix
+==========  =====================================
```
I would remove this table, it's not really relevant. In scikit-learn we mostly use CSR / CSC providing an example of how to convert `SparseDataFrame` to CSR as done below might be enough.

fdroessler replied:
Worked on this and the 80 line length here: https://github.com/scikit-learn/scikit-learn/pull/11305/commits/37380c95de850a668ad5516a65cbc7db97c682a7

rth reviewed (commented):
Thanks, @fdroessler , I think it's a good start. 

Could you please re-format the lines to be less or around 80 character longs. I'll add more comments then (as it's easier to put comments in the right place with shorter lines).

@jorisvandenbossche might also be interested in this.

amueller commented:
maybe also @TomAugspurger? I'm sure he has nothing better to do ;)

amueller commented on doc/interoperability.rst:
```diff
@@ -0,0 +1,10 @@
+.. include:: includes/big_toc_css.rst
```
I would like to see this be more of a first class citizen. I don't know how to do that well, though :-/

fdroessler replied:
Not sure how to do this, help appreciated but will look into it myself too.

fdroessler committed:
keep line length around 80 chars

fdroessler committed:
fix code examples in docstring

amueller commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,202 @@
+>>> iris_pd['target'] = y
+>>>
+>>> from sklearn.model_selection import train_test_split
+>>> train, test = train_test_split(iris_pd.copy(), test_size= 0.3)
```
That is weird. Actually we have to copy again afterwards otherwise pandas will throw a warning....

TomAugspurger replied:
I think we've discussed this before :) https://github.com/pandas-dev/pandas/issues/18801

Probably easiest to leave the copy for now.

fdroessler committed:
note on to_numpy

fdroessler committed:
short description on the problem of supporting pandas

fdroessler committed:
corrected comment on .values and in-memory copy.

fdroessler committed:
change wording on SparseDataFrames

fdroessler committed:
some clean-up

fdroessler commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,202 @@
+
+In general, scikit-learn supports the use of `pandas DataFrames <http://pandas.pydata.org/pandas-docs/stable/>`__ implicitly. However, this implicit support has its conditions and potential pitfalls and some (not all) of these will be outlined below.
+
+Every scikit-learn estimator supports the use of DataFrames which is achieved by obtaining a `numpy array <https://docs.scipy.org/doc/numpy/user/>`__ using the :meth:`.values` property of the DataFrame class. There are several conditions on such an input DataFrame one of which is that the data in the DataFrame columns used by the estimator are of numerical type. Other conditions and pitfalls are  ...
```
Good to know thanks for pointing this out. I added a note about this in this commit https://github.com/scikit-learn/scikit-learn/pull/11305/commits/b06790b08472fdf59742aca2c7d5add43f0933d9

fdroessler commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,202 @@
+==================================================================
+
+In general, scikit-learn supports the use of `pandas DataFrames <http://pandas.pydata.org/pandas-docs/stable/>`__ implicitly. However, this implicit support has its conditions and potential pitfalls and some (not all) of these will be outlined below.
+
```
I think I used most of your words but hopefully this is a start: https://github.com/scikit-learn/scikit-learn/pull/11305/commits/e7501a2f9cab6e1d179fa413c6bafb57af9ccd18

fdroessler commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,202 @@
+  >>> all([is_numeric_dtype(df[x]) for x in df.columns])
+  >>> # False
+
+There are also a variety of classes such as pipelines and model selection processes that will pass a DataFrame along "as is" to the nested estimators (see the next section for an example). However, this is not guaranteed and some of the issues with this are outlined below and sources (where available) of discussions and work-arounds (the latter is provided without guarantee that the workaround wi ...
```
See: https://github.com/scikit-learn/scikit-learn/pull/11305/commits/b3896a342b3e0ba8627c780d04322ebc4f612f4e

fdroessler commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,202 @@
+Pandas in Pandas out
+====================
+
+It is currently not the case, that all estimators in scikit-learn return a DataFrame as an output when provided with a DataFrame as an input. There is an ongoing discussion and drive to improve support on this matter although there are also voices that suggest that in general, scikit-learn should only be expected to work smoothly with numpy arrays and have some basic support for DataFrames.
```
Hmm I think I came across this somewhere and thought it interesting but I agree it does not add much value here. 😅  Removed in https://github.com/scikit-learn/scikit-learn/pull/11305/commits/d42c997fbeee1c97cdca8ffe3c8a61291d096c74

fdroessler commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,202 @@
+
+In general, Sparse datastructures (i.e. DataFrames, Series, Arrays) are memory optimised structures of their standard counterparts. They work on the principle that they contain a lot of NaN, 0, or another repeating value (this can be specified), and as such a lot of memory can be saved, which means one can potentially work with datasets that would otherwise be too large to fit into available memo ...
+
+In Pandas, the sparse datastructrures are: :class:`~pandas.SparseDataFrame`, :class:`~pandas.SparseSeries` and :class:`~pandas.SparseArray`.
```
Reworded it slightly here: https://github.com/scikit-learn/scikit-learn/pull/11305/commits/38d98fc151e0b49f12c3851c52aac2d0bb8cbb5c

fdroessler commented:
Thanks @rth @amueller @TomAugspurger  for the feedback. I added some things and mentioned the commits in the comments as well so it is easier for you all to check. I will keep working on the other sections and the overall thing over the next few days. Further feedback is more than welcome!

fdroessler committed:
:bug: fix doctest :memo: fix links for pandas reference

fdroessler committed:
Merge branch 'master' into pandas_userguide

fdroessler committed:
rephrasing of estimators/pipelines/transformers to primitives for the scope of this section

fdroessler commented:
@jorisvandenbossche would be great if you could have a look at this at some point. Also if someone could point me to fix the Travis issues that would be great.

As far as I understand it, it tries to import pandas but that does not exist in the environment so it can't run the doctest on travis? Can I configure this or do I need to skip this? Also it seems that even when setting the random_states I get different outputs on Travis compared to locally. Is there any way to fix this or should I change/reduce the output in the docstring so that it passes the tests?

fdroessler commented:
@jorisvandenbossche @TomAugspurger
Hi I was just wondering if you had another look at this? Thought it would be useful to get it out there sooner rather than later :)

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+Pandas in Pandas out
+====================
+
+It is currently not the case that all estimators in Scikit-learn return a
```
I think actually *none* of the estimators currently do that?

jorisvandenbossche replied:
Yeah, so cross-validation (`train_test_split` you show below) is an exception, but this is not an estimator. So therefore my confusion.

Personally, I would make the cross-validation a separate section.

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+<class 'pandas.core.frame.DataFrame'>
+
+However, removing some random values from the dataset and using the
+:class:`~sklearn.impute.SimpleImputer` to replace the NaNs returns a NumPy
```
I would maybe show here a simpler example such as StandardScaler (then the introduction of missing values is not needed here, keeping the focus on what we actually want to show), and the imputation example can still be used for the imputation section.

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+Because Scikit-learn transforms DataFrames to NumPy arrays, it should be
+assumed, that all information and benefits of column names is lost and that
+from that point forward, only column order and not column labels stay relevant.
+This can cause problems when e.g. pickling a trained estimator and later
```
this is a problem in general with using new data for predict, not only when having pickled a trained estimator, so I would just scope this problem as "predicting on unseen data" where the new data have a different column order.

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+labels, has those in a different order compared to the original DataFrame.
+Intuitively it might be assumed that because Scikit-learn handles the use of
+DataFrames so smoothly in most cases, the same goes for re-ordering labeled
+DataFrames but this is **not** the case.
```
To make this a bit more succint, I would just say that scikit-learn does not check if the column order is consistent or automatically reorders.

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+
+An example of how this might impact your future prediction can be seen in the
+example given below (original with slight modifications adjusting for current
+API, thanks to `SauceCat <https://github.com/scikit-learn/scikit-learn/issues/7242#issue-173131995>`__).
```
Although it is certainly nice to credit people for examples, I think here it is not needed to include this)

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+                   multi_class='auto', n_jobs=4, penalty='l2', random_state=123,
+                   solver='lbfgs', tol=0.0001, verbose=0, warm_start=False)
+>>>
+>>> prob1 = lg.predict_proba(test[feature_columns_train])
```
can you show the output of `predict` instead of `predict_proba` here? (I assume that will be easier to interpret)

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+example given below (original with slight modifications adjusting for current
+API, thanks to `SauceCat <https://github.com/scikit-learn/scikit-learn/issues/7242#issue-173131995>`__).
+
+>>> # for simplification, consider a very simple case
```
this whole code block needs to be intended (and a `::` before it, as in one of the previous comments)

jorisvandenbossche replied:
also, you can remove this first commented line

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+`issue <https://github.com/scikit-learn/scikit-learn/issues/7242>`__ should be
+watched and used to update this paragraph in the future. A simple and straight-
+forward way of ensuring that column ordering and column labels are the same is
+using something like :meth:`df.loc[:, list of column names]` to enforce the
```
I would do here the same as what you used above in the example (so `df[list of column names]` instead of `loc`).

Also, you can remove the `:meth:`, as this is not a method for which sphinx can create a link to the docstring page.

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+Dealing with missing values
+===========================
+
+As per the glosary, most Scikit-learn estimators do not work with missing
```
```suggestion
As per the glossary, most Scikit-learn estimators do not work with missing
```

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+
+As per the glosary, most Scikit-learn estimators do not work with missing
+values. If they do, NaN is the preferred representation of missing values. For
+more details, see https://scikit-learn.org/stable/glossary.html#term-missing-values.
```
There is a sphinx syntax for doing this, I think ``:term:`missing values` ``

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+
+This is an issue which has vastly improved from pandas version 0.21.1 onwards.
+The conversation from DataFrames has been largely optimised and are much faster
+to convert.
```
I think this sentence is out of scope for the material here (we should try to focus on scikit-learn, not on how things evolved in pandas, unless they are very relevant of course).

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+In Pandas, the main sparse data structures is: :class:`~pandas.SparseArray`.
+Both :class:`~pandas.SparseDataFrame` and :class:`~pandas.SparseSeries` exists
+but don't offer significant advantages over DataFrames where the columns are
+SparseArrays and will be deprecated at some point in the future.
```
I would be a bit more vague here. We don't know the exact plan what we are going to do in pandas. So I would keep to the current situation here, and I think the most important point that scikit-learn users need to know is: scikit-learn does not support sparse pandas structures, by default they will also be converted to numpy arrays (and thus densified). If you want to keep it sparse, you need to convert to a sparse Scipy matrix yourself manually. (is that correct? I personally never used it)

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+sparse matrices or pandas DataFrames.
+
+Dealing with missing values
+===========================
```
It might be good to mention here that the imputation methods now also support non-numeric data with the constant and most frequent strategies, and then link to the impute user guide.

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+learn provides an experimental :class:`~sklearn.compose.ColumnTransformer` API.
+This API (which might change in the future) allows the definition of different
+transformation steps to be applied to different columns in either arrays,
+sparse matrices or pandas DataFrames.
```
I would add a link to the user guide section about this.

jorisvandenbossche commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+
+- https://scikit-learn.org/stable/glossary.html#term-categorical-feature
+- https://scikit-learn.org/stable/modules/preprocessing.html#preprocessing-categorical-features
+- https://github.com/scikit-learn-contrib/sklearn-pandas
```
It would be nice to expand this section, now that OneHotEncoder/OrdinalEncoder support non-numeric data (but very short pointing to the user guide is good)

Can you use sphinx syntax to link to the other section in the docs? (see http://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-arbitrary-locations)

If you have time of course!

fdroessler replied:
I have addressed all of your other comments. I will work on this section over the next days and ping you when it is updated. Have a look at the rest and see if they are ok now.

jorisvandenbossche reviewed (commented):
@fdroessler Thanks a lot for this! 

I added a bunch of comments. Sorry if they are a bit succint here and there, but know this is very much appreciated.

fdroessler commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,299 @@
+  using :meth:`.values`. More details on this can be found in the
+  `release notes <http://pandas-docs.github.io/pandas-docs-travis/whatsnew/v0.24.0.html#accessing-the-values-in-a-series-or-index>`__
+  and the documentation `here <http://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#basics-dtypes>`__
+  and `here <http://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#attributes-and-underlying-data>`__.
```
I am happy to remove the section but as per discussion above I got the feeling that mentioning it in a small note might be worth. After all it could be that some scikit-learn users that are also pandas users will hear about it this way because there are not following all the pandas updates. This could be appreciated by some.

jnothman replied:
I agree this is not relevant.

fdroessler replied:
removed

fdroessler committed:
:memo: update based on Joris comments

fdroessler committed:
adding pandas to osx travis to pass doctests

fdroessler committed:
comment on categorical data

fdroessler committed:
Merge branch 'master' into pandas_userguide

fdroessler committed:
various linkages

fdroessler committed:
include pandas in macos test for doctests

fdroessler commented:
@jorisvandenbossche it would be great if you could have another look at this, I addressed most of your comments, a few are still up for discussion or waiting for a comment. Thanks :)

fdroessler committed:
Merge branch 'master' into pandas_userguide

fdroessler committed:
Introduction sentence

fdroessler commented:
@amueller @TomAugspurger @jorisvandenbossche Sorry to ping you all here but it would be nice to move this along a bit. I was wondering if this is generally what you were expecting or if there a still major things that would need changing before this can be accepted?

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+capability of supporting heterogenous data in one data structure.
+
+Every Scikit-learn estimator/transformer/pipeline
+(for the rest of this section we shall call these primitives)
```
these should just be called estimators

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+
+Every Scikit-learn estimator/transformer/pipeline
+(for the rest of this section we shall call these primitives)
+supports the use of DataFrames as inputs which is achieved by obtaining a
```
```suggestion
supports the use of DataFrames as inputs. Most do this by obtaining a
```

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+(for the rest of this section we shall call these primitives)
+supports the use of DataFrames as inputs which is achieved by obtaining a
+`NumPy array <https://docs.scipy.org/doc/numpy/user/>`__ using
+the :meth:`~numpy.asarray` on a DataFrame object. The only exception where a
```
Not the only exception. Several meta-estimators pass their data on unmodified (Pipeline, etc.) or know how to slice a Pandas DataFrame (GridSearchCV, etc.)

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+  and `here <http://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html#attributes-and-underlying-data>`__.
+
+There are several conditions on using a DataFrame as an input to
+Scikit-learn primitives, one of which is that the data in the
```
```suggestion
most Scikit-learn estimators, one of which is that the data in the
```

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+(where available) of
+discussions and work-arounds (the latter is provided without guarantee that the
+workaround will still work) are provided. It should also be mentioned that if
+the DataFrame contains heterogenous data, the :meth:`~numpy.asarray()` function will
```
even numeric data of heterogenous dtype

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+
+There are also a variety of classes such as pipelines and model selection
+processes that will pass a DataFrame along "as is" to the nested estimators
+(see the next section for an example). However, this is not guaranteed and some
```
I don't know what "this is not guaranteed" means. For particular estimators, this behaviour is guaranteed, insofar as we guarantee anything.

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+Pandas in **not** Pandas out
+============================
+
+Some primitives in Scikit-learn support the pandas in pandas out, however it
```
I think this should focus on output of `transform` and `predict` and similar methods. `train_test_split` is necessarily a special case and can be noted. Apart from FunctionTransformer with appropriate args and Pipeline with appropriate args, I can't think of any estimator that will output a pandas type from `predict`, `transform`, etc.

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+  array([0, 0, 2, 0, 0])
+
+
+At the time of writing, it is the users responsibility to ensure that the
```
```suggestion
At the time of writing, it is the user's responsibility to ensure that the
```

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+At the time of writing, it is the users responsibility to ensure that the
+column ordering in the data used for training the estimator is the same as the
+ordering of the data used for prediction. There is an ongoing discussion
+whether or not this will change in the future and this
```
I think the only likelihood in the future is that where it would otherwise be passed to asarray and its column names mismatch those used at fit, this will lead to an error. You shouldn't give the impression that we might support automatically reordering columns. This is a very unlikely (and risky) behaviour for us to support.

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+=========================
+
+For a general guide on how to get started with categorical features please refer
+to :term:`categocrical feature` and :ref:`preprocessing_categorical_features`.
```
```suggestion
to :term:`categorical feature` and :ref:`preprocessing_categorical_features`.
```

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+Dealing with missing values
+===========================
+
+As per the glossary, most Scikit-learn primitives do not work with missing
```
```suggestion
As per the glossary, most Scikit-learn estimators do not work with missing
```

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,295 @@
+.. note::
+  **Issue:**
+  ``Sparse DataFrames`` are not automatically converted to ``scipy.sparse``
+  matrices.
```
Sparse DataFrames are also deprecated in Pandas, if I'm not mistaken.

fdroessler replied:
I was not able to find evidence, I thought they still exist? https://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html?highlight=sparsedataframe ?

jnothman reviewed (commented):
This is looking pretty comprehensive!

jnothman committed:
Update doc/interoperability/pandas.rst

Co-Authored-By: fdroessler <11539266+fdroessler@users.noreply.github.com>

jnothman committed:
Update doc/interoperability/pandas.rst

Co-Authored-By: fdroessler <11539266+fdroessler@users.noreply.github.com>

jnothman committed:
Update doc/interoperability/pandas.rst

Co-Authored-By: fdroessler <11539266+fdroessler@users.noreply.github.com>

jnothman committed:
Update doc/interoperability/pandas.rst

Co-Authored-By: fdroessler <11539266+fdroessler@users.noreply.github.com>

fdroessler committed:
Update doc/interoperability/pandas.rst

Co-Authored-By: Joel Nothman <joel.nothman@gmail.com>

fdroessler committed:
Update based on jnothman comments

fdroessler committed:
merge remote

fdroessler committed:
Merge branch 'master' into pandas_userguide

fdroessler commented:
Thanks @jnothman for the review, apologies that it took some time to fix these but I had a lot on my plate. I would appreciate if you have time to look over my changes based on your feedback.

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,288 @@
+=======================
+
+This part of the User Guide aims to highlight some of the particularities that
+arise when using Pandas datastructures as input for Scikit-learn.
```
```suggestion
arise when using Pandas data structures as input for Scikit-learn.
```

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,288 @@
+
+2) Either the :meth:`~pandas.SparseDataFrame.to_coo()` method on the Pandas
+SparseDataFrame, or :class:`~scipy.sparse.coo_matrix` constructor are
+alternative ways you can convert to a scipy sparse datastructure.
```
```suggestion
alternative ways you can convert to a scipy.sparse data structure.
```

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,288 @@
+Scikit-learn supports the use of
+`Pandas DataFrames <http://pandas.pydata.org/pandas-docs/stable/>`__
+implicitly. However, this implicit support has its conditions and potential
+pitfalls and some (not all) of these will be outlined below.
```
```suggestion
pitfalls, some of which are outlined below.
```

Your text is generally very verbose, but I won't give it a full copy edit. Perhaps you can look over it for sentences you can say in fewer, simple words and paragraphs you can say in fewer sentences.

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,288 @@
+implicitly. However, this implicit support has its conditions and potential
+pitfalls and some (not all) of these will be outlined below.
+
+Reasons for these conditions and the difficulty of supporting Pandas, stems
```
This sentence could start at "Pandas data structures ..."

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,288 @@
+pitfalls and some (not all) of these will be outlined below.
+
+Reasons for these conditions and the difficulty of supporting Pandas, stems
+from the fact that Pandas data structures such as Series and DataFrames can
```
This does not really apply to Series

jnothman commented on doc/interoperability/pandas.rst:
```diff
@@ -0,0 +1,288 @@
+Reasons for these conditions and the difficulty of supporting Pandas, stems
+from the fact that Pandas data structures such as Series and DataFrames can
+hold heterogenous datatypes (different columns can contain different
+datatypes). Therefore, Series and DataFrames can be thought of as containers
```
"while numpy arrays are designed to work with homogeneous data type, and do not support some rich dtypes available in Pandas, such as..." would summarise the remainder of the paragraph.

lorentzenchr commented:
Much has happened since. We have #29957, see also https://github.com/scikit-learn/scikit-learn/issues/10453#issuecomment-2413799691.
